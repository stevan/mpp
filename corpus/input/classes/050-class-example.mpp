class Point {
    field $x :param :reader;
    field $y :param :reader;

    sub at ($x, $y) {
        return Point->new( x => $x, y => $y )
    }

    method add ($other) {
        return Point->new(
            x => $self->x() + $other->x(),
            y => $self->y() + $other->y(),
        );
    }
}

class Point3D :isa(Point) {
    field $z :param :reader;

    sub at ($x, $y, $z) {
        return Point3D->new( x => $x, y => $y, z => $z )
    }

    method add ($other) {
        return Point3D->new(
            x => $self->x() + $other->x(),
            y => $self->y() + $other->y(),
            z => $self->z() + $other->z(),
        );
    }
}
