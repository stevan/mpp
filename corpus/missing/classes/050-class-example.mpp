    }
    method add ($other) {
        return Point->new(
        );
    }
}
    }
    method add ($other) {
        return Point3D->new(
        );
    }
}