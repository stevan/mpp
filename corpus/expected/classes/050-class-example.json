[
  {
    "type": "Class",
    "name": "Point",
    "body": [
      {
        "type": "Field",
        "variable": {
          "type": "Variable",
          "name": "$x"
        },
        "attributes": [
          "param",
          "reader"
        ]
      },
      {
        "type": "Field",
        "variable": {
          "type": "Variable",
          "name": "$y"
        },
        "attributes": [
          "param",
          "reader"
        ]
      },
      {
        "type": "Sub",
        "name": "at",
        "parameters": [
          {
            "type": "Parameter",
            "variable": {
              "type": "Variable",
              "name": "$x"
            }
          },
          {
            "type": "Parameter",
            "variable": {
              "type": "Variable",
              "name": "$y"
            }
          }
        ],
        "body": [
          {
            "type": "Return",
            "value": {
              "type": "MethodCall",
              "object": {
                "type": "Variable",
                "name": "Point"
              },
              "method": "new",
              "arguments": [
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "x"
                  },
                  "right": {
                    "type": "Variable",
                    "name": "$x"
                  }
                },
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "y"
                  },
                  "right": {
                    "type": "Variable",
                    "name": "$y"
                  }
                }
              ]
            }
          }
        ]
      },
      {
        "type": "Method",
        "name": "add",
        "parameters": [
          {
            "type": "Parameter",
            "variable": {
              "type": "Variable",
              "name": "$other"
            }
          }
        ],
        "body": [
          {
            "type": "Return",
            "value": {
              "type": "MethodCall",
              "object": {
                "type": "Variable",
                "name": "Point"
              },
              "method": "new",
              "arguments": [
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "x"
                  },
                  "right": {
                    "type": "BinaryOp",
                    "operator": "+",
                    "left": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$self"
                      },
                      "method": "x",
                      "arguments": []
                    },
                    "right": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$other"
                      },
                      "method": "x",
                      "arguments": []
                    }
                  }
                },
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "y"
                  },
                  "right": {
                    "type": "BinaryOp",
                    "operator": "+",
                    "left": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$self"
                      },
                      "method": "y",
                      "arguments": []
                    },
                    "right": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$other"
                      },
                      "method": "y",
                      "arguments": []
                    }
                  }
                }
              ]
            }
          }
        ]
      }
    ]
  },
  {
    "type": "Class",
    "name": "Point3D",
    "body": [
      {
        "type": "Field",
        "variable": {
          "type": "Variable",
          "name": "$z"
        },
        "attributes": [
          "param",
          "reader"
        ]
      },
      {
        "type": "Sub",
        "name": "at",
        "parameters": [
          {
            "type": "Parameter",
            "variable": {
              "type": "Variable",
              "name": "$x"
            }
          },
          {
            "type": "Parameter",
            "variable": {
              "type": "Variable",
              "name": "$y"
            }
          },
          {
            "type": "Parameter",
            "variable": {
              "type": "Variable",
              "name": "$z"
            }
          }
        ],
        "body": [
          {
            "type": "Return",
            "value": {
              "type": "MethodCall",
              "object": {
                "type": "Variable",
                "name": "Point3D"
              },
              "method": "new",
              "arguments": [
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "x"
                  },
                  "right": {
                    "type": "Variable",
                    "name": "$x"
                  }
                },
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "y"
                  },
                  "right": {
                    "type": "Variable",
                    "name": "$y"
                  }
                },
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "z"
                  },
                  "right": {
                    "type": "Variable",
                    "name": "$z"
                  }
                }
              ]
            }
          }
        ]
      },
      {
        "type": "Method",
        "name": "add",
        "parameters": [
          {
            "type": "Parameter",
            "variable": {
              "type": "Variable",
              "name": "$other"
            }
          }
        ],
        "body": [
          {
            "type": "Return",
            "value": {
              "type": "MethodCall",
              "object": {
                "type": "Variable",
                "name": "Point3D"
              },
              "method": "new",
              "arguments": [
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "x"
                  },
                  "right": {
                    "type": "BinaryOp",
                    "operator": "+",
                    "left": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$self"
                      },
                      "method": "x",
                      "arguments": []
                    },
                    "right": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$other"
                      },
                      "method": "x",
                      "arguments": []
                    }
                  }
                },
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "y"
                  },
                  "right": {
                    "type": "BinaryOp",
                    "operator": "+",
                    "left": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$self"
                      },
                      "method": "y",
                      "arguments": []
                    },
                    "right": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$other"
                      },
                      "method": "y",
                      "arguments": []
                    }
                  }
                },
                {
                  "type": "BinaryOp",
                  "operator": "=>",
                  "left": {
                    "type": "String",
                    "value": "z"
                  },
                  "right": {
                    "type": "BinaryOp",
                    "operator": "+",
                    "left": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$self"
                      },
                      "method": "z",
                      "arguments": []
                    },
                    "right": {
                      "type": "MethodCall",
                      "object": {
                        "type": "Variable",
                        "name": "$other"
                      },
                      "method": "z",
                      "arguments": []
                    }
                  }
                }
              ]
            }
          }
        ]
      }
    ],
    "parent": "Point"
  }
]